### DISCUSSION TITLE
`GitIndex` does not find latest versions (?)

### DISCUSSION THREAD

**Discussion by MingweiSamuel** [frewsxcv/rust-crates-index#160] [Q&A] [answered]
Says the latest version is 0.3.0 but there is actually also 0.4.0 and 0.5.0. What gives?

https://crates.io/crates/hydroflow/0.5.0

https://github.com/rust-lang/crates.io-index/blob/95a6dad481be5b7893c82d419785036b0dee85fc/hy/dr/hydroflow

## vs
```toml
[dependencies]
crates-index = { version = "2.5.0", features = [ "git" ] }
```
```rust
use crates_index::GitIndex;

fn main() {
    let index = GitIndex::new_cargo_default().unwrap();
    let c = index.crate_("hydroflow").unwrap();
    println!("{:#?}", c.versions().iter().map(|v| v.version()).collect::<Vec<_>>());
}
```
```
[
    "0.0.0",
    "0.0.1",
    "0.0.2",
    "0.1.0",
    "0.1.1",
    "0.2.0",
    "0.3.0",
]
```
https://github.com/MingweiSamuel/crates-index-test

**Comment by MingweiSamuel** [comment]
Only goes up to `0.28.0` when I tested `gitoxide`, should be `0.33.0`

**Comment by Byron** [comment, marked as answer]
You will want to call `update()` to be sure it's the latest version of the Git index.

```Rust
use crates_index::GitIndex;

fn main() {
    let mut index = GitIndex::new_cargo_default().unwrap();
    // ------------------------
    index.update().unwrap();
    // ------------------------
    let c = index.crate_("gix-path").unwrap();
    println!("{:#?}", c.versions().iter().map(|v| v.version()).collect::<Vec<_>>());
}
```

**Reply by MingweiSamuel** [reply]
It doesn't work on my local windows machine. Unable to reproduce on gh-actions though

```rust
        let mut index = GitIndex::new_cargo_default().unwrap();
        println!("URL {}", index.url());
        index.update().unwrap();
        let c = index.crate_("gitoxide").unwrap();
        println!(
            "{:#?}",
            c.versions().iter().map(|v| v.version()).collect::<Vec<_>>()
        );
```

```log
...
    "0.23.0",
    "0.24.0",
    "0.25.0",
    "0.26.0",
    "0.27.0",
    "0.28.0",
]
```

**Reply by MingweiSamuel** [reply]
Is there some sort of cache I can clear somewhere?

**Reply by Byron** [reply]
It's this folder here: `~/.cargo/registry/index/github.com-1ecc6299db9ec823` that could be removed. **However**, before doing that, it would be great if you could share it with me as I suspect a bug that leads to updates not being done properly, which I have seen before. If you could share it with me, that would be appreciated - to my knowledge it doesn't contain any PII as long as you didn't change it by hand.

**Reply by MingweiSamuel** [reply]
That worked. Here is the old cache which doesn't work. It's noticeably larger than the reset/working version

old big version: https://storage.googleapis.com/cmmz/misc/github.com-1ecc6299db9ec823.zip
new working version: https://storage.googleapis.com/cmmz/misc/github.com-1ecc6299db9ec823.new.zip

**Reply by Byron** [reply]
Thanks for sharing, it's very large indeed! It's interesting as `git gc` should remove unreachable objects which there should be many of. But that's another issue for another day.

I will take a look and let you know the results - my hope is to be able to fixup `crates-index` to be able to deal with such an index as well.

Please feel free to remove both files from storage, I downloaded the old version.

**Reply by Byron** [reply]
I could reproduce the issue and the fix is here: https://github.com/frewsxcv/rust-crates-index/pull/161.

The fix was released here: https://github.com/frewsxcv/rust-crates-index/releases/tag/v2.5.1 .

Thanks again for your help - this issue was critical and it seems nobody is really using `crates-index` anymore or else that should have come up already.
